### ISSUE TITLE
Unify categories

### ISSUE THREAD

**Issue by muffinresearch** [mozilla/addons-server#5989]
We've talked about this here and there before and I think the consensus was that unifying the add-on categories in AMO is something we want to do.

To that end do we can we sort out what the mappings should be and discuss what it would take to make the change?

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/ADDSRV-149)

**Comment by jvillalobos** [comment]
Just to clarify scope, we're referring to the categories for extensions on desktop and Android, right?

Other categories, which presumably are out of scope:
* Thunderbird and SeaMonkey: we plan to move to a separate site.
* Dictionaries and Language Packs: they have a single category per application, but that's just an implementation detail for not having categories.
* Themes and Complete Themes: they have their own categories, and the latter are going away with legacy add-ons.
* Search Tools: they have different categories, would be hard to consolidate because they relate more to content than features. Not sure if categories are useful for them.

**Comment by muffinresearch** [comment]
> Just to clarify scope, we're referring to the categories for extensions on desktop and Android, right?

I think themes would be included in this since that's currently part of the issue - see #4766 for more info.

**Comment by eviljeff** [comment]
Also clarifying scope: this should be able unifying categories for the different _applications_, not the different add-on types.  Lang packs, themes, etc, are different types to Extensions, so will have different categories still.

**Comment by jvillalobos** [comment]
Agree that themes (lightweight and future static) should share one set of categories that work both on desktop and Android.

**Comment by eviljeff** [comment]
@jvillalobos can you (or @devaneymoz or as appropriate) provide mappings from the old to new combined categories?

Desktop Firefox has: 
```
'alerts-updates': StaticCategory(id=72, name=_(u'Alerts & Updates')),
'appearance': StaticCategory(id=14, name=_(u'Appearance')),
'bookmarks': StaticCategory(id=22, name=_(u'Bookmarks')),
'download-management': StaticCategory(id=5, name=_(u'Download Management')),
'feeds-news-blogging': StaticCategory(id=1, name=_(u'Feeds, News & Blogging')),
'games-entertainment': StaticCategory(id=142, name=_(u'Games & Entertainment')),
'language-support': StaticCategory(id=37, name=_(u'Language Support')),
'photos-music-videos': StaticCategory(id=38, name=_(u'Photos, Music & Videos')),
'privacy-security': StaticCategory(id=12, name=_(u'Privacy & Security')),
'search-tools': StaticCategory(id=13, name=_(u'Search Tools')),
'shopping': StaticCategory(id=141, name=_(u'Shopping')),
'social-communication': StaticCategory(id=71, name=_(u'Social & Communication')),
'tabs': StaticCategory(id=93, name=_(u'Tabs')),
'web-development': StaticCategory(id=4, name=_(u'Web Development')),
'other': StaticCategory(id=73, name=_(u'Other'))
```
Android Firefox has:
```
'device-features-location': StaticCategory(id=145, name=_(u'Device Features & Location')),
'experimental': StaticCategory(id=151, name=_(u'Experimental')),
'feeds-news-blogging': StaticCategory(id=147, name=_(u'Feeds, News, & Blogging')),
'performance': StaticCategory(id=144, name=_(u'Performance')),
'photos-media': StaticCategory(id=143, name=_(u'Photos & Media')),
'security-privacy': StaticCategory(id=149, name=_(u'Security & Privacy')),
'shopping': StaticCategory(id=150, name=_(u'Shopping')),
'social-networking': StaticCategory(id=148, name=_(u'Social Networking')),
'sports-games': StaticCategory(id=146, name=_(u'Sports & Games')),
'user-interface': StaticCategory(id=152, name=_(u'User Interface')),
'other': StaticCategory(id=153, name=_(u'Other'))
```
(ignoring Seamonkey and Thunderbird)

**Comment by jvillalobos** [comment]
These are all straightforward:

* 'feeds-news-blogging' -> 'feeds-news-blogging'
* 'photos-media' -> 'photos-music-videos'
* 'security-privacy' -> 'privacy-security'
* 'shopping' -> 'shopping'
* 'social-networking' -> 'social-communication'
* 'sports-games' -> 'games-entertainment'
* 'user-interface' -> 'appearance'
* 'other' -> 'other'

These not so much:

* 'device-features-location' -> 'other'
* 'experimental' -> 'other'
* 'performance' -> 'other'

@devaneymoz has final say here, so let's wait for his input.

**Comment by diox** [comment]
Slight detail that will make the migration a bit more painful to write: When an add-on is in the "other" category, it should be its only category.

**Comment by eviljeff** [comment]
We can always have new categories, if none of the existing desktop Firefox ones are appropriate.

**Comment by jvillalobos** [comment]
I don't think any of those 3 are good add-on categories, but again that's not my call :).

**Comment by devaneymoz** [comment]
I support @jvillalobos suggested category mapping. Good idea to collapse those three Android categories (device location, experimental, performance) into Other. 

Separately, there has been expressed desire to retool our desktop/Android categories to modernize the nomenclature and perhaps even collapse a couple of existing categories into each other to streamline the ecosystem. I wrote up a [PRD](https://docs.google.com/document/d/1-5Ad5kcKtwJizyDm2Jrz9hLy9dTYwxUcSHbHkXTKz_0/edit#) some time ago but it was shelved for other priority projects. Anyway I know this would represent a separate project but wanted to mention it in case it impacts this work. Retooling our categories is something I'd like to do in 2018.

**Comment by eviljeff** [comment]
This may be easier to do after we only have to think about Firefox desktop and android.  After November 10th.

**Comment by muffinresearch** [comment]
Marking as blocked on the migration for thunderbird and seamonkey.

**Comment by davidhedlund** [comment]
Should be added to "Add-on Type" in https://addons.mozilla.org/en-US/firefox/search/
* Dictionary
* Language Pack
* Complete Theme

Is this the right issue to discuss this in or should I open a separate issue?

**Comment by jvillalobos** [comment]
It's a separate issue.

**Comment by AlexandraMoga** [comment]
In the context of new add-on submissions, there is a scenario that should also be solved once this is fixed:

1. submit a new add-on compatible with Firefox desktop only - during submission you will select categories specific for desktop
2. submit a new version that is compatible with both desktop and Android - for new versions, category selection is not required

This means that for Android, your add-on will appear 'category-less', unless the user knows that he will have to choose an Android category in the Edit information page, after the new version is submitted.

**Comment by eviljeff** [comment]
I think #5981 may need to be done at the same time as this (or maybe not).  Either way, we'll need confirmation about the new categories so it's blocked on that.  

@AlexandraMoga having add-ons with no categories on Android (or on Desktop, if the platform choices were reversed) is something that needs to be fixed regardless of merging categories (and sooner) so can you file a separate issue for that?

**Comment by jvillalobos** [comment]
A new set of categories is planned for Q1 at the moment, just FYI.

**Comment by eviljeff** [comment]
I guess the planned work to decide on the new set of categories in Q1 2019 didn't happen but I still think it'd help so keeping open for now.

**Comment by eviljeff** [comment]
Note, when/after we do this we should clear up all the old `AddonCategory` instances in the database so that only the instances that reference current category IDs should remain.  (They're harmless because the static category code only return categories in the constant, but redundant if they exist)

**Comment by diox** [comment]
We currently expose (and accept incoming data) categories in the API as an object with app names (`firefox` or `android`) as the keys, and the categories themselves as a list.

With the unification it no longer makes sense. I'm proposing:
- We continue to allow the old format as input in all API versions
- We continue to expose categories per app in v4 and below, but we don't look what the add-on is compatible with, we just expose the same list for `firefox` and `android` at all times
- In v5, we expose categories as a flag list of categories with no app instead of an object and also allow that format as input

v4 format:
```json
"categories": {
  "firefox": ["something", "something-else"],
  "android": ["something", "something-else"],
}
```

v5 format:
```json
"categories": ["something", "something-else"],
```

We'd have to make the frontend compatible first.

**Comment by eviljeff** [comment]
I think ideally v5 would only accept the list style format of categories, and only support the objects with applications as keys for backwards compatibility in v4/3 (e.g. we'd just strip it down to the dict values straight away), but if that's too onerous I suppose it doesn't hurt to support object inputs in v5 (undocumented)

**Comment by diox** [comment]
Category mapping (android -> firefox) copied from JIRA:
- device-features-location —> other
- experimental —> other
- feeds-news-blogging —> feeds-news-blogging
- performance —> other
- photos-media —> photos-music-videos
- security-privacy —> privacy-security
- shopping —> shopping
- social-networking —> social-communication
- sports-games —> games-entertainment
- user-interface —> appearance
- other —> other

**Comment by diox** [comment]
Categories are no longer going to be app-specific (though they remain type-specific: langpacks, dicts and themes have different categories than extensions).

The migration of old Android-specific categories is done through a specific command. Until it has been ran, all extensions that had only android-specific categories would show no categories (and those that had both Android and Firefox ones will only show the Firefox ones). Once migrated, extensions with Android categories will gain new categories according to the mapping in the previous comment.

Addon detail API v5 is changed to return flat categories `"categories": ["foo", "bar"]` where v3/v4 still shows the old structure `"categories": {"firefox": ["foo", "bar"]}` without showing Android categories anymore.

Categories list API v5 no longer returns the `application` field. v4/v3 still does, but again, only `firefox`.

**Comment by jvillalobos** [comment]
🙌🏽

**Comment by ioanarusiczki** [comment]
@diox 

Been looking around with the new submission API, I've only one question:
- at addon creation or addon edit when I'm sending the old format `{"categories":{"firefox":["shopping"], "android":["sports-games"]}}` I won't receive any warning , addon is created or edited and android information is ignored. I hope that's expected.

**Comment by diox** [comment]
Yes, I wasn't so nice as to handle old categories in the API, so the `android` specific ones are ignored. It should be ok.

**Comment by ioanarusiczki** [comment]
@diox 

Migration has been triggered on -dev. 

Some ids got an extra category. An example, 358616 has now 3 categories warning is: "You can have only 2 categories." when I try to save it from dev hub.

Another example, 627575 had "other" with "games and entertainment" -> "The miscellaneous category cannot be combined with additional categories." 

Ids without a category got at least one, mainly the "other".

**Comment by ioanarusiczki** [comment]
@diox I've noted down some ids for -stage so whenever you're ready.

**Comment by diox** [comment]
I ran the command on stage and reindexed.

**Comment by ioanarusiczki** [comment]
@diox 

On stage:

- ids that had 2 categories have 4 now but trying cu save this would throw an error    " You can have only 2 categories."
- ids that did not have any category received other + social communication this would error too from dev hub "    The miscellaneous category cannot be combined with additional categories."

I'll mark this issue verified.

**Comment by KevinMind** [comment]
Old Jira Ticket: https://mozilla-hub.atlassian.net/browse/ADDSRV-149
